File: loans/ios/Classes/ScLoanEventsHandler.swift
import Foundation
import Flutter

class ScLoanEventsHandler: NSObject {
    
    private var eventChannel: FlutterEventChannel?
    private var eventStreamHandler: ScLoanEventsStreamHandler?
    
    func setup(with registrar: FlutterPluginRegistrar) {
        print("Setting up SCLoans events channel...")
        
        eventStreamHandler = ScLoanEventsStreamHandler()
        eventChannel = FlutterEventChannel(name: "scloans_events", binaryMessenger: registrar.messenger())
        eventChannel?.setStreamHandler(eventStreamHandler)
        
        setupNotificationListener()
        print("SCLoans events channel setup complete")
    }
    
    private func setupNotificationListener() {
        NotificationCenter.default.addObserver(
            forName: Notification.Name("scloans_notification"),
            object: nil,
            queue: .main
        ) { [weak self] notification in
            self?.handleNotification(notification)
        }
    }
    
    private func handleNotification(_ notification: Notification) {
        guard let eventStreamHandler = eventStreamHandler,
              let userInfo = notification.userInfo,
              let jsonString = userInfo["payload_str"] as? String else { return }
        
        eventStreamHandler.sendEvent(jsonString)
    }
    
    deinit {
        NotificationCenter.default.removeObserver(self)
        eventChannel?.setStreamHandler(nil)
    }
}

File: loans/android/src/main/kotlin/com/example/scloans/ScLoanEventsHandler.kt
package com.example.scloans

import android.util.Log
import com.smallcase.loans.data.listeners.Notification
import com.smallcase.loans.data.listeners.NotificationCenter
import com.smallcase.loans.core.external.ScLoanNotification
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel


class ScLoanEventsHandler : FlutterPlugin {
    
    private val TAG = "ScLoanEventsHandler"
    private lateinit var eventChannel: EventChannel
    private lateinit var eventStreamHandler: ScLoanEventsStreamHandler
    private var notificationObserver: ((Notification) -> Unit)? = null
    
    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        setup(flutterPluginBinding)
    }
    
    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        cleanup()
    }
    
    private fun setup(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        eventStreamHandler = ScLoanEventsStreamHandler()
        eventChannel = EventChannel(flutterPluginBinding.binaryMessenger, "scloans_events")
        eventChannel.setStreamHandler(eventStreamHandler)
        
        setupNotificationListener()
        Log.d(TAG, "ScLoan events channel setup complete")
    }
    
    private fun setupNotificationListener() {
        notificationObserver?.let { NotificationCenter.removeObserver(it) }
        
        notificationObserver = { notification ->
            try {
                val jsonString = notification.userInfo?.get(ScLoanNotification.STRINGIFIED_PAYLOAD_KEY) as? String
                if (jsonString != null) {
                    Log.d(TAG, "Sending notification to Flutter: $jsonString")
                    eventStreamHandler.sendEvent(jsonString)
                } else {
                    Log.w(TAG, "Notification has no JSON payload")
                }
            } catch (e: Exception) {
                Log.e(TAG, "Error processing notification: ${e.message}")
                eventStreamHandler.sendError("NOTIFICATION_PARSE_ERROR", e.message, e.toString())
            }
        }
        
        notificationObserver?.let { 
            NotificationCenter.addObserver(it)
            Log.d(TAG, "Notification observer added")
        }
    }
    
    private fun cleanup() {
        try {
            notificationObserver?.let { 
                NotificationCenter.removeObserver(it)
                Log.d(TAG, "Notification observer removed")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error removing notification observer: ${e.message}")
        }
        
        notificationObserver = null
        eventChannel.setStreamHandler(null)
        Log.d(TAG, "SCGateway events cleanup complete")
    }
}